### ISSUE TITLE
[tune] tune.run uses methods that are not in ProgressReporter 

### ISSUE THREAD

**Issue by andrew-rosenfeld-ts** [ray-project/ray#13153]
### What is the problem?

While the [docs](https://docs.ray.io/en/master/tune/api_docs/reporters.html) say you can write your own custom `ProgressReporter`, `tune.run()` calls methods that are not in `ProgressReporter`. 

I'm not sure what should change - either `ProgressReporter` needs these methods, or `tune.run() `shouldn't expect them to be there.

This happens with Ray 1.2 nightly as of 2021-01-02. (Python, TensorFlow, OS, etc versions not relevant)

### Reproduction
Here's a repro script that just stitches together the "Quick Start" example from the tune docs with the CustomReporter code from the reporter docs.

```python
from ray import tune
from ray.tune import ProgressReporter


class CustomReporter(ProgressReporter):

    def should_report(self, trials, done=False):
        return True

    def report(self, trials, *sys_info):
        print(*sys_info)
        print("\n".join([str(trial) for trial in trials]))


def objective(step, alpha, beta):
    return (0.1 + alpha * step / 100)**(-1) + beta * 0.1


def training_function(config):
    # Hyperparameters                                                                                                                                                                                              
    alpha, beta = config["alpha"], config["beta"]
    for step in range(10):
        # Iterative training function - can be any arbitrary training procedure.                                                                                                                                   
        intermediate_score = objective(step, alpha, beta)
        # Feed the score back back to Tune.                                                                                                                                                                        
        tune.report(mean_loss=intermediate_score)


tune.run(
    training_function,
    config={
        "alpha": tune.grid_search([0.001, 0.01, 0.1]),
        "beta": tune.choice([1, 2, 3])
    },
    progress_reporter=CustomReporter())

```

results in
```
AttributeError: 'CustomReporter' object has no attribute 'set_search_properties'
```

and I am pretty sure that after fixing this, it will then hit an issue with `set_total_samples()`.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by sumanthratna** [comment]
Hi @andrew-rosenfeld-ts, I think you should be subclassing `TuneReporterBase` and not `ProgressReporter`. Can you try that let me know if it works? If so, we'll update the docs. Thanks!

**Comment by andrew-rosenfeld-ts** [comment]
Hi @sumanthratna , that "works" in the sense that it runs properly, but I don't think it's the right decision from a design perspective. `TuneReporterBase` imposes a fairly particular structure that people might not want to inherit, and then `ProgressReporter` would also not really serve a purpose. I think the solution is probably changing the usage in `tune.run()` (or maybe adding these methods to `ProgressReporter`, but they don't look like "necessary" methods).

**Comment by sumanthratna** [comment]
> `TuneReporterBase` imposes a fairly particular structure that people might not want to inherit, and then `ProgressReporter` would also not really serve a purpose. I think the solution is probably changing the usage in `tune.run()` (or maybe adding these methods to `ProgressReporter`, but they don't look like "necessary" methods).

this makes sense; thoughts @richardliaw?

**Comment by krfricke** [comment]
This was introduced when we switched to passing `metric` and `mode` to `tune.run()`. I think you're making a good point here. I'd argue we will want to  add `set_search_properties` and `set_total_samples` to the `ProgressReporter` base class. While I agree that not everyone extending `ProgressReporter` might need the information passed by these methods, it would be even less desirable to have them extend `TuneReporterBase` if they do. A simple noop method will suffice.

Would you like to file a PR for this @andrew-rosenfeld-ts?

On a different note, this kind of error will be discovered once we add type hinting to `tune.run()`. This should also be straightforward to implement.

**Comment by andrew-rosenfeld-ts** [comment]
@krfricke sure, I can try a PR for this.

**Comment by krfricke** [comment]
We addressed this in #13684 when introducing type hinting to `tune.run()`.

**Comment by andrew-rosenfeld-ts** [comment]
Thanks for taking care of this Kai, and sorry I didn't get to it myself.
